packages/client/lib/sync/fullsync.ts

```diff
@@ -1,34 +1,105 @@
 import { BN } from 'ethereumjs-util'
-import VM from '@ethereumjs/vm'
 import { Peer } from '../net/peer/peer'
 import { short } from '../util'
 import { Synchronizer, SynchronizerOptions } from './sync'
 import { BlockFetcher } from './fetcher'
 import { Block } from '@ethereumjs/block'
+import VM from '@ethereumjs/vm'
+import { DefaultStateManager } from '@ethereumjs/vm/dist/state'
+import { SecureTrie as Trie } from '@ethereumjs/trie'
 
 /**
  * Implements an ethereum full sync synchronizer
  * @memberof module:sync
  */
 export class FullSynchronizer extends Synchronizer {
+  private blockFetcher: BlockFetcher | null
+
   public vm: VM
+  public runningBlocks: boolean
 
-  private blockFetcher: BlockFetcher | null
+  private stopSyncing: boolean
+  private vmPromise?: Promise<void>
+
+  // Tracking vars for log msg condensation on zero tx blocks
+  private NUM_ZERO_TXS_PER_LOG_MSG = 50
+  public zeroTxsBlockLogMsgCounter: number = 0
 
   constructor(options: SynchronizerOptions) {
     super(options)
     this.blockFetcher = null
 
     if (!this.config.vm) {
+      const trie = new Trie(this.stateDB)
+
+      const stateManager = new DefaultStateManager({
+        common: this.config.common,
+        trie,
+      })
+
       this.vm = new VM({
         common: this.config.common,
         blockchain: this.chain.blockchain,
+        stateManager,
       })
     } else {
       this.vm = this.config.vm
       //@ts-ignore blockchain has readonly property
       this.vm.blockchain = this.chain.blockchain
```

Comment:
Just noticing this here, we don't enforce that VM uses the `StateManager` which is attached to the `stateDB`. Not sure if I/we (?) intended this.

Reply:
Yes, noticed this as well at some point and decided not to put too much attention right now, but this is a bit an open field where we need to put a bit more continued thought, not so pressing though I would say.